source/opt/scalar_analysis_simplification.cpp

```diff
@@ -0,0 +1,531 @@
+  if (operand_1->GetType() == SENode::Constant)
+    constant = operand_1;
+  else if (operand_2->GetType() == SENode::Constant)
+    constant = operand_2;
+
+  // If the expression is not a variable multiplied by a constant coefficient,
+  // exit out.
+  if (!(value_unknown && constant)) {
+    return false;
+  }
+
+  int64_t sign = negation ? -1 : 1;
+
+  auto iterator = accumulators_.find(value_unknown);
+  int64_t new_value = constant->AsSEConstantNode()->FoldToSingleValue() * sign;
+  // Add the result of the multiplication to the accumulators.
+  if (iterator != accumulators_.end()) {
+    (*iterator).second += new_value;
+  } else {
+    accumulators_.insert({value_unknown, new_value});
+  }
+
+  return true;
+}
+
+SENode* SENodeSimplifyImpl::Simplify() {
+  // We only handle graphs with an addition, multiplication, or negation, at the
+  // root.
+  if (node_->GetType() != SENode::Add && node_->GetType() != SENode::Multiply &&
+      node_->GetType() != SENode::Negative)
+    return node_;
+
+  SENode* simplified_polynomial = SimplifyPolynomial();
+
+  SERecurrentNode* recurrent_expr = nullptr;
+  node_ = simplified_polynomial;
+
+  // Fold recurrent expressions which are with respect to the same loop into a
+  // single recurrent expression.
+  simplified_polynomial = FoldRecurrentAddExpressions(simplified_polynomial);
+
+  simplified_polynomial =
+      EliminateZeroCoefficientRecurrents(simplified_polynomial);
+
+  // Traverse the immediate children of the new node to find the recurrent
+  // expression. If there is more than one there is nothing further we can do.
+  for (SENode* child : simplified_polynomial->GetChildren()) {
+    if (child->GetType() == SENode::RecurrentAddExpr) {
+      recurrent_expr = child->AsSERecurrentNode();
+    }
+  }
+
+  // We need to count the number of unique recurrent expressions in the DAG to
+  // ensure there is only one.
+  for (auto child_iterator = simplified_polynomial->graph_begin();
+       child_iterator != simplified_polynomial->graph_end(); ++child_iterator) {
+    if (child_iterator->GetType() == SENode::RecurrentAddExpr &&
+        recurrent_expr != child_iterator->AsSERecurrentNode()) {
+      return simplified_polynomial;
+    }
+  }
+
+  if (recurrent_expr) {
+    return SimplifyRecurrentAddExpression(recurrent_expr);
+  }
+
+  return simplified_polynomial;
+}
+
+// Traverse the graph to build up the accumulator objects.
+void SENodeSimplifyImpl::GatherAccumulatorsFromChildNodes(SENode* new_node,
+                                                          SENode* child,
+                                                          bool negation) {
+  int32_t sign = negation ? -1 : 1;
+
+  if (child->GetType() == SENode::Constant) {
+    // Collect all the constants and add them together.
+    constant_accumulator_ +=
+        child->AsSEConstantNode()->FoldToSingleValue() * sign;
+
+  } else if (child->GetType() == SENode::ValueUnknown ||
```

Comment:
Shouldn't ValueUnknown just be added to the graph?  Why are they handled together with RecurrentAddExpr?

Reply:
Here the code is calculating the net sum of each unique `ValueUnknown` or `RecurrentAddExpr` in the graph. To rebuild the graph of `X+X+X*2` into `4*X` we count the occurrences of X and create a new node of `count*X` after. We want that still to be true if X is a recurrent expression or, say, a load. Subsequent simplifications will apply only to `RecurrentAddExpr` nodes but at this point we're treating them as generic variables.

Reply:
Ah, right.  ValueUnknown and CantCompute are different.  Thanks.  Could you put this example in a comment?  Thanks.